Stable and asynchronous KNX library.

Julius2342 opened the pull request:
Overview
--------

The current KNX implementation is not very stable and not asynchronous. The [XKNX](http://xknx.io/) Home Assistant Plugin has the following advantages to the current implementation:

XKNX...
* ... does [cooperative multitasking via asyncio](https://github.com/XKNX/xknx/blob/master/examples/example_light_state.py) and is 100% thread safe.
* ... is a Python library that cleanly exposes the API.
* ... provides support for KNX/IP [routing](https://github.com/XKNX/xknx/blob/master/xknx/io/routing.py) *and* [tunneling](https://github.com/XKNX/xknx/blob/master/xknx/io/tunnel.py) devices.
* ... has strong coverage with [unit tests](https://github.com/XKNX/xknx/tree/master/test).
* ... automatically updates and synchronizes all devices in the background periodically.
* ... listens for all updates of all devices on the KNX bus and updates the corresponding HASS device.
* ... has a clear abstraction of data/network/logic-layer.
* ... provides Heartbeat monitoring for Tunneling connections + clean reconnect if KNX/IP connection failed.
* ... does clean [connect](https://github.com/XKNX/xknx/blob/master/xknx/io/connect.py) and [disconnect](https://github.com/XKNX/xknx/blob/master/xknx/io/disconnect.py) requests to the tunneling device.


Configuration:
--------------

### Platform:

```yaml
knx:
```

Optional, recommended for large KNX installations (>100 devices) and/or if you want to use the XKNX abstraction also for other scripted tools outside HASS:

```yaml
knx:
  config_file: '/path/to/xknx.yaml'
```
* **config_file**: (*Optional*) path for xknx configuration file.

If the autodetection of the KNX/IP device does not work you can specify ip/port of the tunneling device:

```yaml 
knx:
  tunneling:
    host: '192.168.2.23'
    port: 3671
    local_ip: '192.168.2.109'
```
* **host**: Host of KNX/IP tunneling device
* **port**: Port of KNX/IP tunneling device
* **local_ip**: IP of local interface

Explicit connction to a KNX/IP routing device:

```yaml
knx:
  config_file: '/path/to/xknx.yaml'
  routing:
     local_ip: '192.168.2.109'
```
* **local_ip**: local ip of interface (which should be used for multicasting)

```yaml
knx:
    fire_event: True
    fire_event_filter: ["1/0/*", "6/2,3,4-6/*"]
```

* **fire_event** (Optional) if set to True, platform will write all received KNX messages to event bus
* **fire_event_filter** If `fire_event` is set `fire_event_filter` has to be specified. `fire_event_filter` defines a list of pattterns for filtering KNX addresses. Only telegrams which match this pattern are send to the HASS event bus. 

### Light:

```yaml
light:
  - platform: knx
    name: Kitchen-Light-1
    address: '1/0/9'
    brightness_address: '1/0/11'

  - platform: knx
    name: Kitchen-Light-2
    address: '1/0/12'
    brightness_address: '1/0/14'
```
* **name** (*Optional*): A name for this devices used within Home Assistant.
* **address**: KNX group address for switching light on and off
* **brightness_address**: (Optional) KNX group address for dimming light.
* **state_address**: (*Optional*) separate KNX group address for retrieving the switch state of the light
* ***brightness_state_address***: (Optional) separate KNX group address for retrieving the dimm state of the light. 

### Switch:

```yaml
switch:
  - platform: knx
    name: Kitchen.Coffee
    address: '1/1/6'
```
* **name** (*Optional*): A name for this devices used within Home Assistant.
* **address**: KNX group address for switching switch on/off
* **state_address**: (*Optional*) separate KNX group address for retrieving the switch state. 

### Sensor:

```yaml
sensor:
  - platform: knx
    name: Heating.Valve1
    address: '2/0/0'
    type: 'percent'

  - platform: knx
    name: Kitchen.Temperature
    address: '6/2/1'
    value_type: 'temperature'
```

* **name** (*Optional*): A name for this devices used within Home Assistant.
* **address**: KNX group address of sensor
* **value_type**: (Optional) "percent", "temperature", "brightness", "speed_ms", "current"


### Binary Sensor:

```yaml
binary_sensor:
    - platform: knx
      name: "Entrance.Motion.Sensor"
      address: '6/0/2'
      device_class: 'motion'
      #significant_bit: 2
```

* **name** (*Optional*): A name for this devices used within Home Assistant.
* **address**: KNX group address of binary sensor
* **device_class**: (Optional) HASS device class e.g. "motion"
* **significant_bit**: (Optional) Specify which significant bit of the KNX value should be used. Default is 1.

You can also attach actions to binary sensors (e.g. to switch on a light when a switch was pressed). In this example I switch on one light when the button was pressed once and two others when the button was pressed a second time.

```yaml
binary_sensor:
    - platform: knx
      name: Livingroom.3Switch3
      address: '5/0/26'
      automation:
        - counter: 1
          hook: 'on'
          action:
            - entity_id: light.hue_color_lamp_1
              service: homeassistant.turn_on
        - counter: 2
          hook: 'on'
          action:
            - entity_id: light.hue_bloom_1
              service: homeassistant.turn_on
            - entity_id: light.hue_bloom_2
              service: homeassistant.turn_on
```

* **name** (*Optional*): A name for this devices used within Home Assistant.
* **counter**: (*Optional*) Set to 2 if your only want the action to be executed if the button was pressed twice. To 3 for 3 times button pressed. Defaults to 1.
* **hook**: (Optional): Indicates if the automation should be executed on what state of the binary sensor. Values: "on" or "off". Defaults to "on".
* **action**: Specify a list of actions analog to the [HASS automation rules](https://home-assistant.io/docs/automation/action/).

### Cover

```yaml
cover:
    - platform: knx
      name: "Kitchen.Shutter"
      move_long_address: '3/0/0'
      move_short_address: '3/0/1'
      position_address: '3/0/3'
      position_state_address: '3/0/2'
      travelling_time_down: 51
      travelling_time_up: 61
```
* **name** (*Optional*): A name for this devices used within Home Assistant.
* **move_long_address**: KNX group address for moving cover full up or down.
* **move_short_address**: (Optional) KNX group address for moving cover short time up or down.
* **position_address**: (*Optional*) KNX group address for moving cover to dedicated position.
* **position_state_address**: (Optional) Separate KNX group address for requesting the current position of cover.
* **angle_address**: (*Optional*) KNX group address for moving cover to dedicated angle.
* **angle_state_address**: (Optional) Separate KNX group address for requesting the current angle of cover.
* **travelling_time_down**: (*Optional*) Time cover needs to travel down in seconds. Needed to calculate the intermediate positions of cover while traveling. Defaults to 25.
* **travelling_time_up**: (*Optional*) Time cover needs to travel up in seconds. Needed to calculate the intermediate positions of cover while traveling. Defaults to 25.

* **invert_position**: (*Optional*) Set this to true if your actuator report fully closed as 100%
* **invert_angle**: (*Optional*) Set this to true if your actuator reports tilt fully closed as 100%


### Climate:

```yaml
climate:
   - platform: knx
     name: HASS-Kitchen.Temperature
     temperature_address: '6/2/1'
     setpoint_address: '5/1/2'
     target_temperature_address: '5/1/1'
     operation_mode_address: '5/1/3'
     operation_mode_state_address: '5/1/4'
     controller_status_address: '5/1/3'
     controller_status_state_address: '5/1/4'
```

alternatively, if your device has dedicated binary group addresses for frost/night/comfort mode:

```yaml
climate:
   - platform: knx
     name: HASS-Kitchen.Temperature
     temperature_address: '6/2/1'
     setpoint_address: '5/1/2'
     target_temperature_address: '5/1/1'
     operation_mode_frost_protection_address: '5/1/3'
     operation_mode_night_address: '5/1/4'
     operation_mode_comfort_address: '5/1/5'
```


* **name** (*Optional*): A name for this devices used within Home Assistant.
* **temperature_address**: KNX group address for reading current room temperature from KNX bus.
* **target_temperature_address**: KNX group address for reading current target temperature from KNX bus.
* **setpoint_address**: KNX group address for basis setpoint

* **operation_mode_address** (*Optional*) KNX address for operation mode (Frost protection/night/comfort).
* **operation_mode_state_address** (*Optional*) KNX address for reading operation mode
* **controller_status_address** (*Optional*) KNX address for HVAC controller status (in accordance with KNX AN 097/07 rev 3)
* **controller_status_state_address** (*Optional*) KNX address for reading HVAC controller status


* **operation_mode_frost_protection_address** (*Optional*) KNX address for switching on/off frost/heat protection mode.
* **operation_mode_night_address** (*Optional*) KNX address for switching on/off night nmode.
* **operation_mode_comfort_address** (*Optional*) KNX address for switching on/off comfort mode.

`operation_mode_frost_protection_address` / `operation_mode_night_address` / `operation_mode_comfort_address` are not necessary if `operation_mode_address` was specified.

### Notify

```yaml
notify:
    - platform: knx
      name: Alarm
      address: '5/1/10'
```

* **name** (*Optional*): A name for this devices used within Home Assistant.
* **address**: KNX group address of notification


### Service

In order to directly interact with the KNX bus, you can now use the following service:

Domain: knx
Service: send 
Service Data: {"address": "1/0/15", "payload": 0}

* **address** : KNX group addresss
* **payload** : payload, either an integer or an array of integers



Known issues:
-------------

Due to lame multicast support the routing abstraction and the gateway scanner
only work with python >=3.5.

Julius2342 committed:
First draft of XKNX module for Home-Assistant

Julius2342 committed:
XKNX does now take path of xknx.yaml as parameter

Julius2342 committed:
Merge remote-tracking branch 'origin/dev' into xknx

Julius2342 committed:
Merge remote-tracking branch 'origin/dev' into xknx

Julius2342 committed:
small fix, telegram_received_callback has different signature

Julius2342 committed:
Merge remote-tracking branch 'origin/dev' into xknx

Julius2342 committed:
changed method of registering callbacks of devices

Julius2342 committed:
removed non async command lines from xknx

Julius2342 committed:
telegram_received_cb not needed within HASS module

Julius2342 committed:
updated requirements

Julius2342 committed:
Merge remote-tracking branch 'origin/dev' into xknx

Julius2342 committed:
Configuration if XKNX should connect via Routing or Tunneling

Julius2342 committed:
bumping version to 0.6.1

Julius2342 committed:
small fix within xknx plugin

Julius2342 committed:
bumped version

Julius2342 committed:
XKNX-Switches are now BinarySensors and Logic from Sensor was moved to BinarySensor

Julius2342 committed:
renamed Outlet to Switch

Julius2342 committed:
pylint

Julius2342 committed:
configuration of KNX lights via HASS config, yay!

Julius2342 committed:
changed name of attribute

Julius2342 committed:
Added configuration for xknx to switch component

Julius2342 committed:
added support for sensors within hass configuration

Julius2342 committed:
added support for climate within hass configuration

Julius2342 committed:
Thermostat -> Climate

Julius2342 committed:
Merge branch 'dev' into xknx

Julius2342 committed:
added configuration support for binary_sensors

Julius2342 committed:
renamed Shutter to Cover

Julius2342 committed:
added configuration support for cover

Julius2342 committed:
restructured file structure according to HASS requirements

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
updated version

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
pylint

Julius2342 committed:
added setpoint support for climate devices

Julius2342 committed:
devices are now in a different module

Julius2342 committed:
more asyncio :-)

Julius2342 committed:
pydocstyle

Julius2342 committed:
pydocstyle

Julius2342 committed:
added actions to binary_sensor

Julius2342 committed:
allow more than one automation

Julius2342 committed:
readded requirement

Julius2342 committed:
Modifications suggested by hound

Julius2342 committed:
Modifications suggested by hound

Julius2342 committed:
Modifications suggested by hound

Julius2342 committed:
Modifications suggested by hound

Julius2342 committed:
xknx now imported as local import

Julius2342 committed:
hound *sigh*

Julius2342 committed:
lint

elupus commented:
Initial tests look good. Much more performant compared to the other knx component. Especially when you have sensors not responding to reads. Since thos are done in parallel and async in this component that doesn't block other components.

I got several errors on first start, the requirements of installing xknx didn't block component setup. Not sure if that is a hass issue in general or just with this component.

elupus commented:
A feature request for light component. alarm_state. My dimmers have a an alarm function for overtemp/disconnect. Would be nice to get that represented somehow.

Julius2342 commented:
@elupus : Isn't this just a binary sensor?

What should happen if a dimmer gets too hot?

Julius2342 committed:
'fixed' coverage.

Julius2342 committed:
next try for getting gen_requirements_all.py working

Julius2342 committed:
removed blank line

Julius2342 commented:
@elupus : Do you still have the log of the first start up? (when the error occurred?) 

Btw. i started working on https://github.com/XKNX/xknx/issues/46 . My plan is to have three loggers:

* one general for errors etc.
* one for KNX/IP packets
* one for business logic (e.g. light turned on/off)

Does this make sense?

elupus commented:
Nope, don't have it. But it was generic trace logging i saw. It was obvious
it skipped logging module since it didn't have timestamps.

On Jul 31, 2017 13:56, "Julius Mittenzwei" <notifications@github.com> wrote:

> @elupus <https://github.com/elupus> : Do you still have the log of the
> first start up? (when the error occurred?)
>
> Btw. i started working on XKNX/xknx#46
> <https://github.com/XKNX/xknx/issues/46> . My plan is to have three
> loggers:
>
>    - one general for errors etc.
>    - one for KNX/IP packets
>    - one for business logic (e.g. light turned on/off)
>
> Does this make sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/8725#issuecomment-319046542>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdeoGFggEbJftDADjsfJnYxHBtPXT1bks5sTcDMgaJpZM4OnqjS>
> .
>

elupus commented:
I got a hang today which behaves a bit bad. You can't see that it hung from GUI. It accepts commands and update the light state, but nothing is sent on bus.

My connection is a bit flaky, so tunnel goes down from time to time.

```
Jul 31 20:42:49 hass hass[51396]: 2017-07-31 20:42:49 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_executed[L]: service_call_id=140201782009984-42>
Jul 31 20:42:50 hass hass[51396]: Exception while processing telegram: Could not send telegram to tunnel
Jul 31 20:42:50 hass hass[51396]: Traceback (most recent call last):
Jul 31 20:42:50 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 79, in process_telegram
Jul 31 20:42:50 hass hass[51396]:     yield from self.process_telegram_outgoing(telegram)
Jul 31 20:42:50 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 91, in process_telegram_outgoing
Jul 31 20:42:50 hass hass[51396]:     yield from self.xknx.knxip_interface.send_telegram(telegram)
Jul 31 20:42:50 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/knxip_interface.py", line 139, in send_telegram
Jul 31 20:42:50 hass hass[51396]:     yield from self.interface.send_telegram(telegram)
Jul 31 20:42:50 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/tunnel.py", line 111, in send_telegram
Jul 31 20:42:50 hass hass[51396]:     raise Exception("Could not send telegram to tunnel")
Jul 31 20:42:50 hass hass[51396]: Exception: Could not send telegram to tunnel
```

After which all reads seem to give these errors. It doesn't seem to reconnect.
```
Jul 31 04:55:03 hass hass[51396]: Sync Ytterdörr Källare
Jul 31 04:55:03 hass hass[51396]: Sync Regndörr Källare
Jul 31 04:55:03 hass hass[51396]: Sync Uttag Tvättstuga
Jul 31 04:55:04 hass hass[51396]: Could not read value of <Switch name="Uttag Tvättstuga" group_address="<Address str="4/1/51" />" group_address_state="<Address str="4/4/
Jul 31 04:55:04 hass hass[51396]: Sync Uttag Bibliotek
Jul 31 04:55:04 hass hass[51396]: Exception while processing telegram: Could not send telegram to tunnel
Jul 31 04:55:04 hass hass[51396]: Traceback (most recent call last):
Jul 31 04:55:04 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 79, in process_telegram
Jul 31 04:55:04 hass hass[51396]:     yield from self.process_telegram_outgoing(telegram)
Jul 31 04:55:04 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 91, in process_telegram_outgoing
Jul 31 04:55:04 hass hass[51396]:     yield from self.xknx.knxip_interface.send_telegram(telegram)
Jul 31 04:55:04 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/knxip_interface.py", line 139, in send_telegram
Jul 31 04:55:04 hass hass[51396]:     yield from self.interface.send_telegram(telegram)
Jul 31 04:55:04 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/tunnel.py", line 111, in send_telegram
Jul 31 04:55:04 hass hass[51396]:     raise Exception("Could not send telegram to tunnel")
Jul 31 04:55:04 hass hass[51396]: Exception: Could not send telegram to tunnel
Jul 31 04:55:05 hass hass[51396]: Could not read value of <Switch name="Uttag Bibliotek" group_address="<Address str="4/1/65" />" group_address_state="<Address str="4/4/6
Jul 31 04:55:05 hass hass[51396]: Sync Uttag Kök
Jul 31 04:55:05 hass hass[51396]: Exception while processing telegram: Could not send telegram to tunnel
Jul 31 04:55:05 hass hass[51396]: Traceback (most recent call last):
Jul 31 04:55:05 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 79, in process_telegram
Jul 31 04:55:05 hass hass[51396]:     yield from self.process_telegram_outgoing(telegram)
Jul 31 04:55:05 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 91, in process_telegram_outgoing
Jul 31 04:55:05 hass hass[51396]:     yield from self.xknx.knxip_interface.send_telegram(telegram)
Jul 31 04:55:05 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/knxip_interface.py", line 139, in send_telegram
Jul 31 04:55:05 hass hass[51396]:     yield from self.interface.send_telegram(telegram)
Jul 31 04:55:05 hass hass[51396]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/io/tunnel.py", line 111, in send_telegram
Jul 31 04:55:05 hass hass[51396]:     raise Exception("Could not send telegram to tunnel")
Jul 31 04:55:05 hass hass[51396]: Exception: Could not send telegram to tunnel
```


Another interesting thing i notice here is that it seem to be serializing the group reads instead of doing them in parallel. It should be doing.
A: send group read +register listener
B: send group read +register listener
C: send group read +register listener
await any of A,B,C or timeout.

My quick glance of your read function that seemed like what it aught to be doing, but the timestamps on the errors above seem to indicate differently. 2 seconds diff between the error and the attempt to send.

elupus commented:
The reason it updates states looks to be device.py:send(). It awaits the put into the queue. Imho it should await the telegram actually sent on the bus (probably even acked when it's a tunnel)

elupus commented:
I suppose multiple loggers could make sense. But just doing per file with period separated names. ie xknx.devices.device then just using the different log levels for different things you log seems easier to handle.

Julius2342 commented:
Catching up with reality :-)

Looks like your installation is either much bigger or your device is slower then mine. I never experienced these kind of problems in my home (the tunneling is now working for about 6 months).

Will add logging in one of the next days. This should enable us to debug this problem.

Thank you for testing! 

(btw. Are you available on discord or any other IM?)

Julius2342 committed:
XKNX 0.7.1 with logging functionality, replaced some print() calls with _LOGGER

Julius2342 committed:
updated requirements_all.txt

Julius2342 committed:
Merge remote-tracking branch 'origin/dev' into xknx

mag2bue commented:
@Julius2342 It is possible to add an additional ON/OFF switch for climate? It will be great to switch my heaters.

Pls. let me know, how I can support you for testing!

Julius2342 commented:
@mag2bue : what exactly should the module do when changing the operation mode? @resoai had a similar feature request.

resoai commented:
I don't think this is how Climate should work. It should have Comfort/Away/Frost protect modes along with setpoint shift for temperature. KNX should than decide what heating or cooling appliances to utilise.

mag2bue commented:
@Julius2342 @resoai 
I'm using a MDT controller. To switch Comfort/Away/Frostt  is it for each a 1bit switch. In my case, I use only comfort ON/OFF

Julius2342 commented:
regarding the climate thing, i think we should move this to a separate discussion: https://github.com/XKNX/xknx/issues/48

resoai commented:
@mag2bue @Julius2342 please have a look at 2.1.3.1 Room Thermostat (page 18): 

http://www.knx.org/fileadmin/template/documents/downloads_support_menu/KNX_tutor_seminar_page/Advanced_documentation/Heating_E1211a.pdf

Julius2342 commented:
@elupus : i opened two tickets for your 2 problems:

* https://github.com/XKNX/xknx/issues/49 Errors on startup
* https://github.com/XKNX/xknx/issues/50 Timout problems on large KNX environments

elupus commented:
Ok. I Will check them. Its not only a timeout issue thou. After the
connection is back, the tunnel doesn't seem to reconnect at all.

On Aug 2, 2017 10:05, "Julius Mittenzwei" <notifications@github.com> wrote:

> @elupus <https://github.com/elupus> : i opened two tickets for your 2
> problems:
>
>    - XKNX/xknx#49 <https://github.com/XKNX/xknx/issues/49> Errors on
>    startup
>    - XKNX/xknx#50 <https://github.com/XKNX/xknx/issues/50> Timout
>    problems on large KNX environments
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/8725#issuecomment-319599512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdeoFArv-sAPyBAd3KREOhg6zduTJYiks5sUC3BgaJpZM4OnqjS>
> .
>

Julius2342 committed:
Merge branch 'dev' into xknx

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

FireFrei commented:
Is it possible to use xknx to access "raw knx messages" in automation like you can do in the current knx component? 
Example: You can create an automation that listens on the event "knx_event" which includes all data received on the knx bus.
This makes it possible to use "special" knx devices that aren't implemented yet. In my case: I have some wall buttons - when I press them, i'd like to run an automation.

Aaaand what about the other way around: Is it possible to send "raw" messages to the knx bus using a HASS service? 

Are these two things already possible? I wasn't able to find a documentation for this purpose :-)

Cheers for your work!

Julius2342 commented:
@FireFrei : for buttons on the wall: this should work already with binary_sensor platform, see example above. (and tell me if something is not working as expected).

and the other way round: what kind of devices would you like to control? If this is not yet possible with current devices we should add some.

FireFrei commented:
@Julius2342 

> for buttons on the wall: this should work already with binary_sensor platform, see example above. (and tell me if something is not working as expected).
Okay, i'll give it a try!

> and the other way round: what kind of devices would you like to control? If this is not yet possible with current devices we should add some.
Like my "Busch-Jäger Tastsensoren" :-) The element has buttons with LEDs. Buttons send to a group address and the LED can change their state/color when they hear a specific message to a specific group address. 

At all, I think adding a service to send data to a group address and the possibility to add listeners to a group address enables your component, to make "custom setups" or special, vendor-specific functions of a device usable :-)

> Update: Made a local test, is this the way you would expect an event:
Yes, exactly :-)

Julius2342 commented:
My touch sensors also have LEDs, i added them as simple lights to hass.

But yeah, i agree, will think of such a service.

Julius2342 committed:
Fixes issue https://github.com/XKNX/xknx/issues/51

FireFrei commented:
I've got another stupid question :-)
The 'primary' group address config field for lights, sensors, binary_sensors is called "address":

> address: KNX group address for switching light on and off

But for switches its "group_address", according to your docs. And switches are called lights in the explanation btw, Is this wanted? :-)
> group_address: KNX group address for switching light on/off

resoai commented:
Perhaps it could be a good idea to keep naming as close to the original knx component as possible in order for people to be able to quickly switch between them?

Julius2342 commented:
@FireFrei : this was just an error within documentation. Should be fixed now.
@resoai : I agree. At least I have to make it backwards compatible.

Julius2342 committed:
https://github.com/XKNX/xknx/issues/52 added raw access to KNX bus from HASS component.

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
bumped version - 0.7.3 contains some bugfixes

Julius2342 committed:
bumped version - 0.7.3 contains some bugfixes

Julius2342 committed:
setting setpoint within climate device has to be async

Julius2342 committed:
bumped version to 0.7.4

Julius2342 committed:
bumped version

resoai commented:
I have also joined testing! I have one suggestion for backward compatibility, let's change "value_type" of the sensor to "type". What do you think?

Julius2342 committed:
https://github.com/XKNX/xknx/issues/48 Adding HVAC support.

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
pylint suggestions

Julius2342 committed:
Made target temperature and set point required attributes

Julius2342 committed:
renamed value_type to type within sensor configuration

Julius2342 committed:
Issue https://github.com/XKNX/xknx/issues/52 : added filter functionality for not flooding the event bus.

Julius2342 committed:
suggestions by pylint

Julius2342 commented:
Current testing update
----------------------

- Added HVAC support as suggested by @resoai  and @mag2bue
  (https://github.com/XKNX/xknx/issues/48)
- Added raw access to KNX bus as suggested by @FireFrei
  (https://github.com/XKNX/xknx/issues/52)
- fixed a bug within cover support reported by @twistermcneal
  (https://github.com/XKNX/xknx/issues/51)
- added retry mechanism according to KNX standard as reported by @elupus
  (https://github.com/XKNX/xknx/issues/50)
- fixed `import` problem as reported by @elupus
  (https://github.com/XKNX/xknx/issues/49)
- added python logging support as suggested by @elupus and @daBONDi
  (https://github.com/XKNX/xknx/issues/46)
- fixed several bugs reported by @andreasnanko

Julius2342 committed:
Added notify support for knx platform.

Julius2342 committed:
logging error if discovery_info is None.

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
review suggestions by @armills

Julius2342 committed:
line too long

pvizeli commented on homeassistant/components/binary_sensor/xknx.py:
```diff
@@ -0,0 +1,164 @@
+
+
+@asyncio.coroutine
+def add_devices_from_component(hass, add_devices):
```
Need not be a coroutine. Only if you Need `yield from` inside a function, that need to be a coroutine. If a function only run inside loop add `@callback`. Add also `async_` sufix to function like this.

armills replied:
@Julius2342 `add_devices_from_component` is being called on the async thread, from `async_setup_platform`. Please make the changes noted by @pvizeli.

pvizeli commented on homeassistant/components/binary_sensor/xknx.py:
```diff
@@ -0,0 +1,164 @@
+    for device in hass.data[DATA_XKNX].xknx.devices:
+        import xknx
+        if isinstance(device, xknx.devices.BinarySensor) and \
+                not hasattr(device, "already_added_to_hass"):
```
That look very wired. Change the software desine to avoid this.

pvizeli commented on homeassistant/components/binary_sensor/xknx.py:
```diff
@@ -0,0 +1,164 @@
+        device_class=config.get(CONF_DEVICE_CLASS),
+        significant_bit=config.get(CONF_SIGNIFICANT_BIT))
+
+    class XKNXAutomation(xknx.devices.ActionBase):
```
Also wired. You generate a class on every time you call `add_devices_from_platform` and mess you memory. Look into `homematic` how are a component / platform like this should look and work.

pvizeli commented:
You need replace the old knx component/platform. Rename it to `knx`

Julius2342 committed:
Using discovery_info to notifiy component which devices should be added.

Julius2342 committed:
moved XKNX automation to main level.

Julius2342 committed:
renamed xknx component to knx.

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
reverted change within .coveragerc

Julius2342 committed:
changed dependency

Julius2342 committed:
updated docstrings.

Julius2342 committed:
updated version of xknx within requirements_all.txt

Julius2342 committed:
moved requirement to correct position

Julius2342 committed:
renamed configuration attribute

Julius2342 commented:
@pvizeli : I implemented the requested changes.

I did not add an `@callback` to `add_devices_from_component` instead i removed the `async` decorator. `add_devices_from_component` is also not called from eventloop ...

Thank you :)

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
added @callback-decorator and async_prefix.

Julius2342 committed:
added @callback decorator and async_ prefix to register_callbacks functions

Julius2342 committed:
Merge branch 'xknx' of github.com:XKNX/home-assistant into xknx

Julius2342 commented:
@pvizeli : @armills helped me to understand the @callback issue. I think now i have implemented all requested changes.

andreasnanko commented:
Using xknx since couple of months now. Looking forward to see it as core homeassistant plugin, no need for me to package as plugin anymore :-)

Julius2342 committed:
Merge remote-tracking branch 'upstream/dev' into xknx

Julius2342 committed:
fixed typo

FireFrei commented:
@andreasnanko I can agree. I'm using xknx for weeks now and its way more stable than the current knx implementation. Works without any issues out of the box. Would appreciate to see it in the next release :-)

Julius2342 committed:
pylint suggestions

Julius2342 committed:
added angle position and invert_position and invert_angle to cover.knx

Julius2342 committed:
typo

Julius2342 committed:
bumped version within requirements_all.txt

Julius2342 committed:
bumped version

elupus commented:
I just grabbed the latest version of this pull request and i get problems with brightness all of a sudden. This used to work. Would be nice of the CouldNotParseTelegram exception actually included some information as to what it is failing to parse :)

```
Sep 05 18:51:05 hass hass[62778]: 2017-09-05 18:51:05 ERROR (MainThread) [xknx.log] Exception while processing telegram: <CouldNotParseTelegram description="" />
Sep 05 18:51:05 hass hass[62778]: Traceback (most recent call last):
Sep 05 18:51:05 hass hass[62778]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 99, in process_telegram
Sep 05 18:51:05 hass hass[62778]:     yield from self.process_telegram_incoming(telegram)
Sep 05 18:51:05 hass hass[62778]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/core/telegram_queue.py", line 128, in process_telegram_incoming
Sep 05 18:51:05 hass hass[62778]:     yield from device.process(telegram)
Sep 05 18:51:05 hass hass[62778]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/devices/light.py", line 170, in process
Sep 05 18:51:05 hass hass[62778]:     yield from self._process_brightness(telegram)
Sep 05 18:51:05 hass hass[62778]:   File "/home/homeassistant/evn_dev/lib/python3.5/site-packages/xknx/devices/light.py", line 177, in _process_brightness
Sep 05 18:51:05 hass hass[62778]:     raise CouldNotParseTelegram()
Sep 05 18:51:05 hass hass[62778]: xknx.exceptions.exception.CouldNotParseTelegram: <CouldNotParseTelegram description="" />
```

Julius2342 commented:
@elupus : Yes, it looks like it tries to parse a normal switch-state address as a brightness address. Will add some information to the exception.

elupus commented:
I have an idea. Maybe you don't filter read requests on the bus? I have openhab running also and it's polling state periodically. So it will issue a read for a state address.

Julius2342 commented:
Ah! This was indeed changed recently! I will try to fix!

elupus commented:
Also this: https://github.com/XKNX/xknx/blob/52d0a4ae43f24c474b93fa7e52b9f64c973ad271/xknx/devices/light.py#L167 looks invalid. 

The operator precedence of python would make the "and" evaluate first. You'd need parentesis there. Doubt that is the issue for me. But a bug non the less.

Julius2342 commented:
13b6bff9d5af2ff8bae7855a268982806b0182c2 fixes first problem
46dbd5334ff326d8089b8d520563e5f4390b685e fixes second problem

elupus commented:
Seems to work fine 👍

TinusK commented:
Works perfekt.

Julius2342 committed:
Added support for HVAC controller status

balloob commented:
:tada: 🐬

Julius2342 commented:
Suggestion for the blog post:
------------------------------

This release ships a new KNX implementation. It will instantly show all changed states of KNX devices within Home Assistant. Additionally it brings support for HVAC devices and notification services. It also adds a service for direct communication with the KNX bus. You can connect to KNX/IP routing and tunnelling devices. In the background it uses asyncio communication.

https://www.youtube.com/watch?v=Fe3yaflU2XM

https://www.youtube.com/watch?v=JI0VJzlGpx4
